File: dotnet/PgStore/PgStore.cs
using System.Text.Json;
using Npgsql;

namespace PgStore;

/// <summary>
/// Shared PostgreSQL-backed store used identically by all .NET web framework benchmarks.
/// This ensures the DB layer is byte-for-byte identical across frameworks.
/// </summary>
public sealed class PgStore : IAsyncDisposable
{
    private readonly NpgsqlDataSource _dataSource;

    public PgStore(string connectionString)
    {
        // Convert PostgreSQL URI format (postgres://user:pass@host:port/db) to ADO.NET format
        if (connectionString.StartsWith("postgres://") || connectionString.StartsWith("postgresql://"))
        {
            var uri = new Uri(connectionString);
            var userInfo = uri.UserInfo.Split(':');
            connectionString = $"Host={uri.Host};Port={uri.Port};Database={uri.AbsolutePath.TrimStart('/')};Username={userInfo[0]};Password={Uri.UnescapeDataString(userInfo.Length > 1 ? userInfo[1] : "")}";
        }

        var builder = new NpgsqlDataSourceBuilder(connectionString);
        builder.ConnectionStringBuilder.MaxPoolSize = 50;
        builder.ConnectionStringBuilder.MinPoolSize = 10;
        _dataSource = builder.Build();
    }

    public async ValueTask DisposeAsync()
    {
        await _dataSource.DisposeAsync();
    }

    /// <summary>Creates the orders + profiles tables and index if they don't exist.</summary>
    public async Task InitSchemaAsync(CancellationToken ct = default)
    {
        await using var cmd = _dataSource.CreateCommand("""
            CREATE TABLE IF NOT EXISTS orders (
                id         BIGSERIAL PRIMARY KEY,
                user_id    VARCHAR(255) NOT NULL,
                status     VARCHAR(50) NOT NULL DEFAULT 'created',
                items      JSONB NOT NULL DEFAULT '[]',
                total      NUMERIC(12,2) NOT NULL DEFAULT 0,
                currency   VARCHAR(10) NOT NULL DEFAULT 'USD',
                created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
            );
            CREATE INDEX IF NOT EXISTS idx_orders_user_id ON orders(user_id);

            CREATE TABLE IF NOT EXISTS profiles (
                user_id  VARCHAR(255) PRIMARY KEY,
                data     JSONB NOT NULL DEFAULT '{}'
            );
            """);
        await cmd.ExecuteNonQueryAsync(ct);
    }

    /// <summary>Truncates both tables. Called before each benchmark run for a clean state.</summary>
    public async Task TruncateAsync(CancellationToken ct = default)
    {
        await using var cmd = _dataSource.CreateCommand(
            "TRUNCATE TABLE orders RESTART IDENTITY; TRUNCATE TABLE profiles");
        await cmd.ExecuteNonQueryAsync(ct);
    }

    // ── CRUD Operations ──

    public async Task<Order> CreateOrderAsync(string userId, List<OrderItem> items, string currency, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(currency)) currency = "USD";

        var total = 0.0;
        foreach (var item in items)
            total += item.Price * item.Quantity;

        var itemsJson = JsonSerializer.Serialize(items);

        await using var cmd = _dataSource.CreateCommand(
            "INSERT INTO orders (user_id, status, items, total, currency) VALUES ($1, 'created', $2, $3, $4) RETURNING id");
        cmd.Parameters.AddWithValue(userId);
        cmd.Parameters.AddWithValue(NpgsqlTypes.NpgsqlDbType.Jsonb, itemsJson);
        cmd.Parameters.AddWithValue(total);
        cmd.Parameters.AddWithValue(currency);

        var id = (long)(await cmd.ExecuteScalarAsync(ct))!;

        return new Order
        {
            OrderId = id.ToString(),
            UserId = userId,
            Status = "created",
            Items = items,
            Total = total,
            Currency = currency
        };
    }

    public async Task<Order?> GetOrderAsync(string userId, string orderId, CancellationToken ct = default)
    {
        await using var cmd = _dataSource.CreateCommand(
            "SELECT id, user_id, status, items, total, currency FROM orders WHERE id = $1 AND user_id = $2");
        cmd.Parameters.AddWithValue(long.Parse(orderId));
        cmd.Parameters.AddWithValue(userId);

        await using var reader = await cmd.ExecuteReaderAsync(ct);
        if (!await reader.ReadAsync(ct))
            return null;

        var itemsJson = reader.GetString(3);
        return new Order
        {
            OrderId = reader.GetInt64(0).ToString(),
            UserId = reader.GetString(1),
            Status = reader.GetString(2),
            Items = JsonSerializer.Deserialize<List<OrderItem>>(itemsJson) ?? [],
            Total = reader.GetDouble(4),
            Currency = reader.GetString(5)
        };
    }

    public async Task<Order?> UpdateOrderAsync(string userId, string orderId, List<OrderItem> items, string currency, CancellationToken ct = default)
    {
        if (string.IsNullOrEmpty(currency)) currency = "USD";

        var total = 0.0;
        foreach (var item in items)
            total += item.Price * item.Quantity;

        var itemsJson = JsonSerializer.Serialize(items);

        await using var cmd = _dataSource.CreateCommand(
            "UPDATE orders SET status = 'updated', items = $1, total = $2, currency = $3 WHERE id = $4 AND user_id = $5");
        cmd.Parameters.AddWithValue(NpgsqlTypes.NpgsqlDbType.Jsonb, itemsJson);
        cmd.Parameters.AddWithValue(total);
        cmd.Parameters.AddWithValue(currency);
        cmd.Parameters.AddWithValue(long.Parse(orderId));
        cmd.Parameters.AddWithValue(userId);

        var rows = await cmd.ExecuteNonQueryAsync(ct);
        if (rows == 0)
            return null;

        return new Order
        {
            OrderId = orderId,
            UserId = userId,
            Status = "updated",
            Items = items,
            Total = total,
            Currency = currency
        };
    }

    public async Task<bool> DeleteOrderAsync(string userId, string orderId, CancellationToken ct = default)
    {
        await using var cmd = _dataSource.CreateCommand(
            "DELETE FROM orders WHERE id = $1 AND user_id = $2");
        cmd.Parameters.AddWithValue(long.Parse(orderId));
        cmd.Parameters.AddWithValue(userId);

        var rows = await cmd.ExecuteNonQueryAsync(ct);
        return rows > 0;
    }

    // ── Bulk & List Operations ──

    public async Task<(List<Order> Orders, double TotalSum)> BulkCreateOrdersAsync(
        string userId, List<BulkOrderInput> orders, CancellationToken ct = default)
    {
        await using var conn = await _dataSource.OpenConnectionAsync(ct);
        await using var tx = await conn.BeginTransactionAsync(ct);

        var results = new List<Order>();
        var totalSum = 0.0;

        foreach (var req in orders)
        {
            var currency = string.IsNullOrEmpty(req.Currency) ? "USD" : req.Currency;
            var total = 0.0;
            foreach (var item in req.Items)
                total += item.Price * item.Quantity;

            var itemsJson = JsonSerializer.Serialize(req.Items);

            await using var cmd = new NpgsqlCommand(
                "INSERT INTO orders (user_id, status, items, total, currency) VALUES ($1, 'created', $2, $3, $4) RETURNING id",
                conn, tx);
            cmd.Parameters.AddWithValue(userId);
            cmd.Parameters.AddWithValue(NpgsqlTypes.NpgsqlDbType.Jsonb, itemsJson);
            cmd.Parameters.AddWithValue(total);
            cmd.Parameters.AddWithValue(currency);

            var id = (long)(await cmd.ExecuteScalarAsync(ct))!;

            results.Add(new Order
            {
                OrderId = id.ToString(),
                UserId = userId,
                Status = "created",
                Items = req.Items,
                Total = total,
                Currency = currency
            });
            totalSum += total;
        }

        await tx.CommitAsync(ct);
        return (results, totalSum);
    }

    public async Task<List<Order>> ListOrdersAsync(string userId, CancellationToken ct = default)
    {
        await using var cmd = _dataSource.CreateCommand(
            "SELECT id, user_id, status, items, total, currency FROM orders WHERE user_id = $1 ORDER BY id");
        cmd.Parameters.AddWithValue(userId);

        await using var reader = await cmd.ExecuteReaderAsync(ct);
        var orders = new List<Order>();

        while (await reader.ReadAsync(ct))
        {
            var itemsJson = reader.GetString(3);
            orders.Add(new Order
            {
                OrderId = reader.GetInt64(0).ToString(),
                UserId = reader.GetString(1),
                Status = reader.GetString(2),
                Items = JsonSerializer.Deserialize<List<OrderItem>>(itemsJson) ?? [],
                Total = reader.GetDouble(4),
                Currency = reader.GetString(5)
            });
        }

        return orders;
    }

    // ── Profile Operations ──

    public async Task UpsertProfileAsync(string userId, Profile profile, CancellationToken ct = default)
    {
        profile.UserId = userId;
        var data = JsonSerializer.Serialize(profile);

        await using var cmd = _dataSource.CreateCommand(
            "INSERT INTO profiles (user_id, data) VALUES ($1, $2) ON CONFLICT (user_id) DO UPDATE SET data = $2");
        cmd.Parameters.AddWithValue(userId);
        cmd.Parameters.AddWithValue(NpgsqlTypes.NpgsqlDbType.Jsonb, data);

        await cmd.ExecuteNonQueryAsync(ct);
    }

    public async Task<Profile?> GetProfileAsync(string userId, CancellationToken ct = default)
    {
        await using var cmd = _dataSource.CreateCommand(
            "SELECT data FROM profiles WHERE user_id = $1");
        cmd.Parameters.AddWithValue(userId);

        await using var reader = await cmd.ExecuteReaderAsync(ct);
        if (!await reader.ReadAsync(ct))
            return null;

        var data = reader.GetString(0);
        return JsonSerializer.Deserialize<Profile>(data);
    }
}

/// <summary>Input type for bulk order creation.</summary>
public class BulkOrderInput
{
    public List<OrderItem> Items { get; set; } = [];
    public string Currency { get; set; } = "";
}

File: java/pgstore/src/main/java/benchmark/pgstore/PgStore.java
package benchmark.pgstore;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.zaxxer.hikari.HikariConfig;
import com.zaxxer.hikari.HikariDataSource;

import java.io.IOException;
import java.net.URI;
import java.sql.*;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * Shared PostgreSQL-backed store used identically by all Java web framework benchmarks.
 * Mirrors the Go pgstore package byte-for-byte in SQL and behaviour.
 */
public class PgStore implements AutoCloseable {

    private static final ObjectMapper MAPPER = new ObjectMapper();
    private final HikariDataSource dataSource;

    /**
     * Creates a PgStore connected to the given PostgreSQL DSN.
     * DSN format: postgres://user:pass@host:port/dbname?sslmode=disable
     * Pool settings: max 50 conns, min 10.
     */
    public PgStore(String dsn) {
        HikariConfig config = new HikariConfig();
        config.setJdbcUrl(dsnToJdbcUrl(dsn));
        config.setMaximumPoolSize(50);
        config.setMinimumIdle(10);
        config.setDriverClassName("org.postgresql.Driver");

        // Extract user/password from DSN
        try {
            URI uri = new URI(dsn.replace("postgres://", "http://").replace("postgresql://", "http://"));
            String userInfo = uri.getUserInfo();
            if (userInfo != null) {
                String[] parts = userInfo.split(":", 2);
                config.setUsername(parts[0]);
                if (parts.length > 1) {
                    config.setPassword(parts[1]);
                }
            }
        } catch (Exception e) {
            throw new RuntimeException("pgstore: parse DSN: " + e.getMessage(), e);
        }

        this.dataSource = new HikariDataSource(config);

        // Ping
        try (Connection conn = dataSource.getConnection()) {
            conn.isValid(5);
        } catch (SQLException e) {
            dataSource.close();
            throw new RuntimeException("pgstore: ping: " + e.getMessage(), e);
        }
    }

    /**
     * Creates the orders + profiles tables and index if they don't exist.
     */
    public synchronized void initSchema() {
        try (Connection conn = dataSource.getConnection();
             Statement stmt = conn.createStatement()) {
            stmt.execute("""
                CREATE TABLE IF NOT EXISTS orders (
                    id         BIGSERIAL PRIMARY KEY,
                    user_id    VARCHAR(255) NOT NULL,
                    status     VARCHAR(50) NOT NULL DEFAULT 'created',
                    items      JSONB NOT NULL DEFAULT '[]',
                    total      NUMERIC(12,2) NOT NULL DEFAULT 0,
                    currency   VARCHAR(10) NOT NULL DEFAULT 'USD',
                    created_at TIMESTAMPTZ NOT NULL DEFAULT NOW()
                );
                CREATE INDEX IF NOT EXISTS idx_orders_user_id ON orders(user_id);

                CREATE TABLE IF NOT EXISTS profiles (
                    user_id  VARCHAR(255) PRIMARY KEY,
                    data     JSONB NOT NULL DEFAULT '{}'
                );
            """);
        } catch (SQLException e) {
            throw new RuntimeException("pgstore: initSchema: " + e.getMessage(), e);
        }
    }

    /**
     * Truncates both tables. Called before each benchmark run for a clean state.
     */
    public void truncate() {
        try (Connection conn = dataSource.getConnection();
             Statement stmt = conn.createStatement()) {
            stmt.execute("TRUNCATE TABLE orders RESTART IDENTITY; TRUNCATE TABLE profiles");
        } catch (SQLException e) {
            throw new RuntimeException("pgstore: truncate: " + e.getMessage(), e);
        }
    }

    // -- CRUD Operations --

    /**
     * Inserts a new order and returns it with the generated ID.
     */
    public Order createOrder(String userId, List<OrderItem> items, String currency) {
        if (currency == null || currency.isBlank()) {
            currency = "USD";
        }

        double total = 0;
        for (OrderItem item : items) {
            total += item.price() * item.quantity();
        }

        String itemsJson;
        try {
            itemsJson = MAPPER.writeValueAsString(items);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("pgstore: marshal items: " + e.getMessage(), e);
        }

        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "INSERT INTO orders (user_id, status, items, total, currency) VALUES (?, 'created', ?::jsonb, ?, ?) RETURNING id")) {
            ps.setString(1, userId);
            ps.setString(2, itemsJson);
            ps.setDouble(3, total);
            ps.setString(4, currency);

            try (ResultSet rs = ps.executeQuery()) {
                rs.next();
                long id = rs.getLong(1);
                return new Order(String.valueOf(id), userId, "created", items, total, currency);
            }
        } catch (SQLException e) {
            throw new RuntimeException("pgstore: insert: " + e.getMessage(), e);
        }
    }

    /**
     * Retrieves an order by user ID and order ID.
     */
    public Optional<Order> getOrder(String userId, String orderId) {
        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "SELECT id, user_id, status, items, total, currency FROM orders WHERE id = ? AND user_id = ?")) {
            ps.setLong(1, Long.parseLong(orderId));
            ps.setString(2, userId);

            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) {
                    return Optional.empty();
                }
                String id = String.valueOf(rs.getLong("id"));
                String uid = rs.getString("user_id");
                String status = rs.getString("status");
                String itemsJson = rs.getString("items");
                double total = rs.getDouble("total");
                String cur = rs.getString("currency");

                List<OrderItem> items = MAPPER.readValue(itemsJson,
                    MAPPER.getTypeFactory().constructCollectionType(List.class, OrderItem.class));
                return Optional.of(new Order(id, uid, status, items, total, cur));
            }
        } catch (SQLException | IOException e) {
            throw new RuntimeException("pgstore: get: " + e.getMessage(), e);
        }
    }

    /**
     * Updates an existing order.
     */
    public Optional<Order> updateOrder(String userId, String orderId, List<OrderItem> items, String currency) {
        if (currency == null || currency.isBlank()) {
            currency = "USD";
        }

        double total = 0;
        for (OrderItem item : items) {
            total += item.price() * item.quantity();
        }

        String itemsJson;
        try {
            itemsJson = MAPPER.writeValueAsString(items);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("pgstore: marshal items: " + e.getMessage(), e);
        }

        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "UPDATE orders SET status = 'updated', items = ?::jsonb, total = ?, currency = ? WHERE id = ? AND user_id = ?")) {
            ps.setString(1, itemsJson);
            ps.setDouble(2, total);
            ps.setString(3, currency);
            ps.setLong(4, Long.parseLong(orderId));
            ps.setString(5, userId);

            int rows = ps.executeUpdate();
            if (rows == 0) {
                return Optional.empty();
            }

            return Optional.of(new Order(orderId, userId, "updated", items, total, currency));
        } catch (SQLException e) {
            throw new RuntimeException("pgstore: update: " + e.getMessage(), e);
        }
    }

    /**
     * Removes an order and returns true if it existed.
     */
    public boolean deleteOrder(String userId, String orderId) {
        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "DELETE FROM orders WHERE id = ? AND user_id = ?")) {
            ps.setLong(1, Long.parseLong(orderId));
            ps.setString(2, userId);

            return ps.executeUpdate() > 0;
        } catch (SQLException e) {
            throw new RuntimeException("pgstore: delete: " + e.getMessage(), e);
        }
    }

    // -- Bulk & List Operations --

    /**
     * Inserts multiple orders in a single transaction.
     */
    public BulkResult bulkCreateOrders(String userId, List<BulkOrderInput> orders) {
        Connection conn = null;
        try {
            conn = dataSource.getConnection();
            conn.setAutoCommit(false);

            List<Order> results = new ArrayList<>();
            double totalSum = 0;

            try (PreparedStatement ps = conn.prepareStatement(
                     "INSERT INTO orders (user_id, status, items, total, currency) VALUES (?, 'created', ?::jsonb, ?, ?) RETURNING id")) {

                for (BulkOrderInput req : orders) {
                    String currency = (req.currency() == null || req.currency().isBlank()) ? "USD" : req.currency();
                    double total = 0;
                    for (OrderItem item : req.items()) {
                        total += item.price() * item.quantity();
                    }

                    String itemsJson = MAPPER.writeValueAsString(req.items());

                    ps.setString(1, userId);
                    ps.setString(2, itemsJson);
                    ps.setDouble(3, total);
                    ps.setString(4, currency);

                    try (ResultSet rs = ps.executeQuery()) {
                        rs.next();
                        long id = rs.getLong(1);
                        results.add(new Order(String.valueOf(id), userId, "created", req.items(), total, currency));
                    }
                    totalSum += total;
                }
            }

            conn.commit();
            return new BulkResult(results, totalSum);
        } catch (Exception e) {
            if (conn != null) {
                try { conn.rollback(); } catch (SQLException ignored) {}
            }
            throw new RuntimeException("pgstore: bulk insert: " + e.getMessage(), e);
        } finally {
            if (conn != null) {
                try { conn.setAutoCommit(true); conn.close(); } catch (SQLException ignored) {}
            }
        }
    }

    /**
     * Returns all orders for a given user.
     */
    public List<Order> listOrders(String userId) {
        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "SELECT id, user_id, status, items, total, currency FROM orders WHERE user_id = ? ORDER BY id")) {
            ps.setString(1, userId);

            List<Order> orders = new ArrayList<>();
            try (ResultSet rs = ps.executeQuery()) {
                while (rs.next()) {
                    String id = String.valueOf(rs.getLong("id"));
                    String uid = rs.getString("user_id");
                    String status = rs.getString("status");
                    String itemsJson = rs.getString("items");
                    double total = rs.getDouble("total");
                    String currency = rs.getString("currency");

                    List<OrderItem> items = MAPPER.readValue(itemsJson,
                        MAPPER.getTypeFactory().constructCollectionType(List.class, OrderItem.class));
                    orders.add(new Order(id, uid, status, items, total, currency));
                }
            }
            return orders;
        } catch (SQLException | IOException e) {
            throw new RuntimeException("pgstore: list: " + e.getMessage(), e);
        }
    }

    // -- Profile Operations --

    /**
     * Creates or updates a user profile (stored as JSONB).
     */
    public void upsertProfile(String userId, Profile profile) {
        Profile withUserId = new Profile(userId, profile.name(), profile.email(), profile.phone(),
            profile.address(), profile.preferences(), profile.paymentMethods(),
            profile.tags(), profile.metadata());

        String data;
        try {
            data = MAPPER.writeValueAsString(withUserId);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("pgstore: marshal profile: " + e.getMessage(), e);
        }

        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "INSERT INTO profiles (user_id, data) VALUES (?, ?::jsonb) ON CONFLICT (user_id) DO UPDATE SET data = ?::jsonb")) {
            ps.setString(1, userId);
            ps.setString(2, data);
            ps.setString(3, data);
            ps.executeUpdate();
        } catch (SQLException e) {
            throw new RuntimeException("pgstore: upsert profile: " + e.getMessage(), e);
        }
    }

    /**
     * Retrieves a user profile.
     */
    public Optional<Profile> getProfile(String userId) {
        try (Connection conn = dataSource.getConnection();
             PreparedStatement ps = conn.prepareStatement(
                 "SELECT data FROM profiles WHERE user_id = ?")) {
            ps.setString(1, userId);

            try (ResultSet rs = ps.executeQuery()) {
                if (!rs.next()) {
                    return Optional.empty();
                }
                String data = rs.getString("data");
                Profile profile = MAPPER.readValue(data, Profile.class);
                return Optional.of(profile);
            }
        } catch (SQLException | IOException e) {
            throw new RuntimeException("pgstore: get profile: " + e.getMessage(), e);
        }
    }

    @Override
    public void close() {
        dataSource.close();
    }

    // -- Utility --

    private static String dsnToJdbcUrl(String dsn) {
        // Convert postgres://user:pass@host:port/db?params to jdbc:postgresql://host:port/db?params
        try {
            URI uri = new URI(dsn.replace("postgres://", "http://").replace("postgresql://", "http://"));
            String host = uri.getHost();
            int port = uri.getPort() > 0 ? uri.getPort() : 5432;
            String path = uri.getPath(); // /dbname
            String query = uri.getQuery();
            String jdbcUrl = "jdbc:postgresql://" + host + ":" + port + path;
            if (query != null && !query.isEmpty()) {
                jdbcUrl += "?" + query;
            }
            return jdbcUrl;
        } catch (Exception e) {
            throw new RuntimeException("pgstore: invalid DSN: " + dsn, e);
        }
    }
}
